File: hw3/task2_1.scala
import org.apache.spark.broadcast.Broadcast
import org.apache.spark.{SparkConf, SparkContext}

import scala.math._
import java.io.PrintWriter

object task2_1 {
  def mergeD[T](x: Set[T], y: Set[T]): Set[T] = {
    x ++ y
  }
  
  def mergeDict[K, V](x: Map[K, V], y: Map[K, V]): Map[K, V] = {
    x ++ y
  }

  // transform weight based on given logic
  def transformWeight(w: Double, nj: Int): Double = {
    val absW = math.abs(w)
    if (nj <= 2) absW else math.pow(absW, 2)
  }

  // inverse frequency
  def inverseFrequency(n: Int, nj: Int): Double = {
    math.log(n.toDouble / (nj + 1))
  }
  def itembase(use: String, bus: String, uniUse: Set[String], uniBus: Set[String], avgStarU: Map[String, Double], avgStarB: Map[String, Double], useDict: Map[String, Set[String]], busDict: Map[String, Set[String]], orgData: Map[String, Map[String, Double]]): Double = {
    if (!uniUse.contains(use)) return 3
    if (!uniBus.contains(bus)) return avgStarU.getOrElse(use, 3.0)  // Users average or fallback

    val n = uniUse.size  // the number of users for inverse frequency calculation
    val otherBus = useDict.getOrElse(use, Set.empty).filter(line => line != bus)

    val busrating = otherBus.flatMap { bu =>
      val combUse = busDict.getOrElse(bus, Set.empty).intersect(busDict.getOrElse(bu, Set.empty))
      val nj = combUse.size
      val fj = inverseFrequency(n, nj)
      //get the value of inverseFrequency
      val w = (nj match {
        case 0 => 1 - abs(avgStarB.getOrElse(bus, 3.0) - avgStarB.getOrElse(bu, 3.0)) / 2
        case 1 => 1 - abs(orgData.getOrElse(bus, Map.empty).getOrElse(combUse.head, 3.0) - avgStarB.getOrElse(bus, 3.0) - orgData.getOrElse(bu, Map.empty).getOrElse(combUse.head, 3.0) + avgStarB.getOrElse(bu, 3.0)) / 5
        case 2 =>
          val diffs = combUse.map { user =>
            val rI = orgData.getOrElse(bu, Map.empty).getOrElse(user, 3.0) - avgStarU.getOrElse(user, 3.0)
            val rJ = orgData.getOrElse(bus, Map.empty).getOrElse(user, 3.0) - avgStarU.getOrElse(user, 3.0)
            abs(rI - rJ)
          }
          val avgDiff = diffs.sum / diffs.size
          1 - avgDiff / 5  // normalize the result
        case _ => {
          val diffs = combUse.map { user =>
            //calculate the diff for same user diff item
            val ri = orgData(bu)(user) - avgStarB(bu)
            val rj = orgData(bus)(user) - avgStarB(bus)
            val nu = ri * fj * rj * fj
            val denI = pow(ri * fj, 2)
            val denJ = pow(ri * fj, 2)
            (nu, denI, denJ)
          }.foldLeft((0.0, 0.0, 0.0)) { case ((num, di, dj), (nu, denI, denJ)) =>
            (num + nu, di + denI, dj + denJ)
          }
          val den = sqrt(diffs._2) * sqrt(diffs._3)
          if (den != 0) diffs._1 / den else 0
        }
      })
      //transferom the weight
      val wTransformed = transformWeight(w, nj)
      val r = orgData.getOrElse(bu, Map.empty).getOrElse(use, avgStarU.getOrElse(use, 3.0))
      Some((wTransformed, r))
    }

    val (nu, den) = busrating.foldLeft((0.0, 0.0)) { case ((nuc, dec), (w, r)) =>
      (nuc + w * r, dec + abs(w))
    }

    if (den != 0) nu / den else 3.0
  }

  def main(args: Array[String]): Unit = {
    // Check command-line arguments
    if (args.length != 3) {
      println("Usage: scala MainApp <train_file_name> <test_file_name> <output_filepath>")
      System.exit(1)
    }

    val train = args(0)
    val test = args(1)
    val output = args(2)
    //val train = "src/main/resources/yelp_train.csv"
    //val test = "src/main/resources/yelp_val.csv"
    //val output = "src/main/resources/output.csv"

    // Initialize SparkContext
    val conf = new SparkConf().setAppName("Task2_1")
    val sc = new SparkContext(conf)

    // Set log level to ERROR to reduce log output
    sc.setLogLevel("ERROR")

    val startTime = System.currentTimeMillis()

    // load training data
    val trainData = sc.textFile(train)
    val headerTrain = trainData.first()
    val data_nh = trainData.filter(line => line != headerTrain)
    val data_rdd_b = data_nh.map(line => {
      val parts = line.split(',')
      (parts(1), (parts(2).toFloat, 1))
    })
    //combine business and stars calculation
    val comb_star_b = data_rdd_b.reduceByKey((a, b) => (a._1 + b._1, a._2 + b._2))
    val avg_star_b = comb_star_b.mapValues { case (sum, count) => sum / count }.collectAsMap()
    // simlar to the business, we user with stars calculation
    val data_rdd_u = data_nh.map(line => {
      val parts = line.split(',')
      (parts(0), (parts(2).toFloat, 1))
    })

    val comb_star_u = data_rdd_u.reduceByKey((a, b) => (a._1 + b._1, a._2 + b._2))
    val avg_star_u = comb_star_u.mapValues { case (sum, count) => sum / count }.collectAsMap()
    //avg_star_u.foreach(println)
    val avg_star_b_broadcast = sc.broadcast(avg_star_b.mapValues(_.toDouble).toMap)
    val avg_star_u_broadcast = sc.broadcast(avg_star_u.mapValues(_.toDouble).toMap)

    val orgData = data_nh.map(line => {
      val parts = line.split(',')
      (parts(1), Map(parts(0) -> parts(2).toDouble))
    }).reduceByKey(mergeDict).collectAsMap().toMap

    val orgDataB = sc.broadcast(orgData)

    val busDict = data_nh.map(line => {
      val parts = line.split(',')
      (parts(1), Set(parts(0)))
    }).reduceByKey(mergeD).collectAsMap().toMap
    val busDictB = sc.broadcast(busDict)

    val useDict = data_nh.map(line => {
      val parts = line.split(',')
      (parts(0), Set(parts(1)))
    }).reduceByKey(mergeD).collectAsMap().toMap
    val useDictB = sc.broadcast(useDict)
    
    val vals = sc.textFile(test)
    val header = vals.first()
    val clean_val = vals.filter(line => line != header).map(row => {
      val p = row.split(",")
      (p(0), p(1))
    })
    val uniUse = useDictB.value.keys.toSet
    val uniBus = busDictB.value.keys.toSet

    val cal = clean_val.map { case (user, business) =>
      ((user, business), itembase(user, business, uniUse, uniBus, avg_star_u_broadcast.value, avg_star_b_broadcast.value, useDictB.value, busDictB.value, orgDataB.value))
    }
    val results = cal.map { case ((user_id, business_id), stars) =>
      s"$user_id,$business_id,$stars"
    }

    val collectedResults = results.collect()
    //output the file into csv
    val writer = new PrintWriter(output)
    try {
      writer.println("user_id,business_id,prediction")
      collectedResults.foreach { resultString =>
        writer.println(resultString)
      }
    } finally {
      writer.close()
    }

    val endTime = System.currentTimeMillis()
    val duration = (endTime - startTime) / 1000.0
    println(s"Duration: $duration")

    sc.stop()
  }

}

File: hw3/task2_1.py
import sys
from pyspark import SparkContext
import os
import time
import math

def merge_d(x, y):
    return x.union(y)
def merge_dict(x, y):
    for k, v in y.items():
        x[k] = v
    return x


def transform_weight(w, nj):
    w = abs(w)
    if nj <= 2:
        return w
    return w ** 2

def inverse_frequency(n, nj):
    return math.log(n / (nj + 1))


def itembase(use, bus):
    if use not in uni_use:
        return 3
    if bus not in uni_bus:
        return avg_star_u.value.get(use, 3)
    ws = []
    rs = []
    n = len(uni_use)
    for bu in use_dict.value[use]:
        if bu == bus:
            continue

        combuse = list(bus_dict.value[bus].intersection(bus_dict.value[bu]))
        nj = len(combuse)
        fj = inverse_frequency(n, nj)
        if nj == 0:
            w = 1 - abs(- avg_star_b.value[bus] + avg_star_b.value[bu]) / 2
        elif nj == 1:
            w = 1 - abs(org_data.value[bus].get(str(combuse), 3) - avg_star_b.value[bus]
                        - org_data.value[bu].get(str(combuse), 3) + avg_star_b.value[bu] ) / 5
      
        elif nj == 2:
          
            diffs = [(org_data.value[bus].get(user, 3) - avg_star_u.value[user]) - (
                        org_data.value[bu].get(user, 3) - avg_star_u.value[user]) for user in combuse]
            avg_diff = sum(abs(diff) for diff in diffs) / 2
            w = 1 - avg_diff / 5  # normalize the result

        else:
            nu = 0.0
            den_i = 0.0
            den_j = 0.0
            for user in combuse:
                r_i = org_data.value[bu].get(user) - avg_star_b.value.get(bu)
                r_j = org_data.value[bus].get(user) - avg_star_b.value.get(bus)
                nu += r_i *fj * r_j*fj
                den_i += (r_i*fj ) ** 2
                den_j += (r_j*fj ) ** 2

            den = (den_i ** 0.5) * (den_j ** 0.5)
            w = (nu / den) if den != 0 else 0
          
        w_transformed = transform_weight(w, nj)
        r = org_data.value[bu].get(use, avg_star_u.value.get(use, 3))
        rs.append(r)
        ws.append(w_transformed)

    nu = sum(w * r for w, r in zip(ws, rs))
    #print(nu)
    den = sum(abs(w) for w in ws)

    rat = nu / den if den != 0 else 3
    return rat


if __name__ == "__main__":

    #check the validation of the command
    if len(sys.argv) != 4:
        print("There is an error for the Usage: python3 task1.py <train_file_name> <test_file_name> <output_filepath>")
        sys.exit(1)
    sc = SparkContext(appName="Task2_1")
    train = sys.argv[1]
    test = sys.argv[2]
    output = sys.argv[3]

    start_time = time.time()

    sc.setLogLevel("ERROR")
    data = sc.textFile(train)
    header = data.first()
    data_nh = data.filter(lambda line: line != header)
    data_rdd_b = data_nh.map(lambda line: (line.split(',')[1], (float(line.split(',')[2]),1)))
    comb_star_b = data_rdd_b.reduceByKey(lambda a, b: (a[0] + b[0], a[1] + b[1]))
    avg_star_b = comb_star_b.mapValues(lambda x: x[0] / x[1]).collectAsMap()

    data_rdd_u = data_nh.map(lambda line: (line.split(',')[0], (float(line.split(',')[2]), 1)))
    comb_star_u = data_rdd_u.reduceByKey(lambda a, b: (a[0] + b[0], a[1] + b[1]))
    avg_star_u = comb_star_u.mapValues(lambda x: x[0] / x[1]).collectAsMap()
    avg_star_b = sc.broadcast(avg_star_b)
    avg_star_u = sc.broadcast(avg_star_u)
    org_data = data_nh.map(lambda line: (line.split(',')[1], {line.split(',')[0]: float(line.split(',')[2])})) \
                .reduceByKey(lambda a, b: merge_dict(a, b)).collectAsMap()
    org_data = sc.broadcast(org_data)

    b_rdd = data_nh.map(lambda line: (line.split(',')[1], {line.split(',')[0]}))
    bus_dict = b_rdd.reduceByKey(lambda a, b: merge_d(a, b)).collectAsMap()
    bus_dict = sc.broadcast(bus_dict)
    u_rdd = data_nh.map(lambda line: (line.split(',')[0], {line.split(',')[1]}))
    use_dict = u_rdd.reduceByKey(merge_d).collectAsMap()
    use_dict = sc.broadcast(use_dict)
    val = sc.textFile(test)
    header_val = val.first()
    clean_val = val.filter(lambda row: row != header_val).map(lambda row: row.split(",")).map(lambda row: (row[0], row[1]))
    uni_use = list(use_dict.value.keys())
    uni_bus = list(bus_dict.value.keys())
    cal = clean_val.map(lambda x: (x, itembase(x[0], x[1])))
    results = cal.map(lambda x: ','.join([str(x[0][0]), str(x[0][1]), str(x[1])]))
    collected_results = results.collect()

    with open(output, 'w') as file:
        file.write('user_id,business_id,prediction\n')
        for line in collected_results:
            file.write(f'{line}\n')

    end_time = time.time()
    execution = end_time - start_time
    print(f"Duration: {execution}")
    sc.stop()
